## 47.1 에러 처리의 필요성

- 에러가 발생하지 않는 코드를 작성하는 것은 불가능함
- 발생한 에러를 방치하면 프로그램은 강제 종료된다

## 47.2 try … catch … finally문

```jsx
try {
  // 실행할 코드(에러가 발생할 가능성이 있는 코드)
} catch (err) {
  // try 코드 블록에서 에러가 발생하면 이 코드 블록의 코드가 실행된다.
  // err 에는 try 코드 브록에서 발생한 Error 객체가 전달된다
} finally {
  // 에러 발생과 상관없이 반드시 한 번 실행된다
}
```

## 47.3 Error 객체

- Error 생성자 함수는 에러 객체를 생성한다
- Error 생성자 함수에는 에러를 상세히 설명하는 에러 메시지를 인수로 전달할 수 있다

```jsx
const error = new Error("invalid");
```

- 자바스크립트는 7가지의 에러 객체를 생성할 수 있는 Error 생성자 함수를 제공한다
- Error 생성자 함수가 생성한 에러 객체의 프로토타입은 모두 Error.prototype을 상속받는다

| 생성자 함수    | 인스턴스                                                                       |
| -------------- | ------------------------------------------------------------------------------ |
| Error          | 일반적인 에러 객체                                                             |
| SyntaxError    | 자바스크립트 문법에 맞지 않는 문을 해석할 때 발생하는 에러 객체                |
| ReferenceError | 참조할 수 없는 식별자를 참조했을 때 발생하는 에러 객체                         |
| TypeError      | 피연산자 또는 인수의 데이터 타입이 유효하지 않을 때 발생하는 에러 객체         |
| RangeError     | 숫자값의 허용 범위를 벗어났을 때 발생하는 에러 객체                            |
| URIError       | encodeURI 또는 decodeURI 함수에 부적절한 인수를 전달했을 때 발생하는 에러 객체 |
| EvalError      | eval 함수에서 발생하는 에러 객체                                               |

## 47.4 throw 문

- 에러를 발생시키려면 try 코드 블록에서 throw 문으로 에러 객체를 던져야 한다
- throw 문의 표현식은 어떤 값이라도 상관없지만 일반적으로 에러 객체를 지정한다

## 47.5 에러의 전파

- 콜 스택의 아래 방향(실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다
- 비동기 함수인 setTimeout 이나 프로미스 후속 처리 메서드의 콜백 함수는 태스크 큐나 마이크로태스크 큐에 일시 저장되었다가 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다. 이때 콜 스택에 푸시된 콜백 함수의 실행 컨텍스트는 콜 스택의 가장 하부에 존재하게 된다. 따라서 에러를 전파할 호출자가 존재하지 않는다.
